### PR TITLE
Check workflow task after reapply events

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
Check workflow task after reapply events

**Why?**
When reapply event to current branch, we need to check if it should schedule a new workflow task

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: service/history/nDCEventsReapplier.go [hunk 1]
```diff
@@ -108,5 +108,18 @@ func (r *nDCEventsReapplierImpl) reapplyEvents(
 		deDupResource := definition.NewEventReappliedID(runID, event.GetEventId(), event.GetVersion())
 		msBuilder.UpdateDuplicatedResource(deDupResource)
 	}
+
+	// After reapply event, checking if we should schedule a workflow task
+	// Do not create workflow task when the workflow is cron and the cron has not been started yet
+	if msBuilder.GetExecutionInfo().CronSchedule != "" && !msBuilder.HasProcessedOrPendingWorkflowTask() {
```
Comment (yiminc) [inline service/history/nDCEventsReapplier.go:114]: instead of checking for cronSchedule, you should check firstWorkflowTaskBackoff

**Review by wxing1292 — APPROVED** [review]

File: service/history/nDCEventsReapplier.go (hunk 1 above)
Comment (yux0) [reply service/history/nDCEventsReapplier.go:114]: If cron has the first wf task, then we should continue schedule a task. Is this correct?

File: service/history/nDCEventsReapplier.go [hunk 2]
```diff
@@ -108,5 +108,18 @@ func (r *nDCEventsReapplierImpl) reapplyEvents(
...
+		return reappliedEvents, nil
+	}
+	if !msBuilder.HasPendingWorkflowTask() {
```
Comment (yycptt) [inline service/history/nDCEventsReapplier.go:117]: Also check if `len(reappliedEvents) != 0`?

File: service/history/nDCEventsReapplier.go (hunk 1 above)
Comment (yycptt) [reply service/history/nDCEventsReapplier.go:114]: We also need to take care of workflow retry backoff case I think.

File: service/history/historyEngine.go [hunk 3]
```diff
@@ -2844,8 +2844,10 @@ func (e *historyEngineImpl) ReapplyEvents(
 				Noop:               false,
 				CreateWorkflowTask: true,
 			}
-			// Do not create workflow task when the workflow is cron and the cron has not been started yet
-			if mutableState.GetExecutionInfo().CronSchedule != "" && !mutableState.HasProcessedOrPendingWorkflowTask() {
+			executionInfo := mutableState.GetExecutionInfo()
+			// Do not create workflow task when the workflow has first workflow task backoff and execution is not started yet
+			workflowTaskBackoff := timestamp.TimeValue(executionInfo.GetExecutionTime()).After(timestamp.TimeValue(executionInfo.GetStartTime()))
+			if workflowTaskBackoff && !mutableState.HasProcessedOrPendingWorkflowTask() {
```
Comment (yycptt) [inline service/history/historyEngine.go:2850]: Shall we create a util function for this check?

**Review by yycptt — APPROVED** [review]

File: service/history/historyEngine.go (hunk 3 above)
Comment (yiminc) [reply service/history/historyEngine.go:2850]: 👍 this is also needed here https://github.com/temporalio/temporal/blob/87c3bebdb4e2a61e6763c62194b33e3967f16dee/service/history/historyEngine.go#L1804

Comment (yux0) [reply service/history/historyEngine.go:2850]: will do

**Review by yiminc — APPROVED** [review]

**Pull request merged**